### DISCUSSION TITLE
关于proxy config里的openai key和发送请求时自带的openai key

### DISCUSSION THREAD

**Discussion by edward-ng-duke** [KenyonY/openai-forward#86] [Q&A] [answered]
想请教您一个问题：
如果发送请求时，自带了一个openai 的key，同时proxy config里也指定了openai 的key，实际效果是使用哪个key？感谢回答 ：）

**Comment by KenyonY** [comment, marked as answer]
hi @83521166 ,
这块的逻辑是这样的：
1. 如果在环境变量中设置了`OPENAI_API_KEY`，同时也设置了 `FORWARD_KEY`, 
会判断请求的key是否在`FORWARD_KEY`中，
   如果在, 则从`OPENAI_API_KEY`中的拿一个key；
   如果不在，则使用自身 (这里应该对应了你的描述中的情况）
2. 如果在环境变量中设置了`OPENAI_API_KEY`，但没有设置`FORWARD_KEY` （不推荐）
   将无视请求中的key, 始终使用配置中的key

**Reply by edward-ng-duke** [reply]
感谢回复！
逻辑很清楚，但是openai现在也不支持分key来查看用量，所以设置FORWARD_KEY的意义是啥？

**Reply by KenyonY** [reply]
FORWARD_KEY 目前的基本作用是保护openai的原始api-key或accessToken不被直接暴露在客户端，这样即使他人拿到FORWARD_KEY也必须在你的转发服务上才能使用；
当然，它还可以用来分发给不同用户，并为每个FORWARD_KEY设置速率限制、Token统计等功能，不过这些功能是企业需要的，对于个人用户不太重要，因此本项目暂未实现
